1) Import Libraries

In [42]:
import pandas as pd
import os
from pathlib import Path

2. Load the dataset

In [43]:
DATA_DIR = Path("../data/raw/fashion-product-images-small")
df = pd.read_csv("../data/processed/styles_clean.csv")
df.head()

,id,gender,masterCategory,subCategory,articleType,baseColour,season,year,usage,productDisplayName
0,15970,Men,Apparel,Topwear,Shirts,Navy Blue,Fall,2011.0,Casual,Turtle Check Men Navy Blue Shirt
1,39386,Men,Apparel,Bottomwear,Jeans,Blue,Summer,2012.0,Casual,Peter England Men Party Blue Jeans
2,59263,Women,Accessories,Watches,Watches,Silver,Winter,2016.0,Casual,Titan Women Silver Watch
3,21379,Men,Apparel,Bottomwear,Track Pants,Black,Fall,2011.0,Casual,Manchester United Men Solid Black Track Pants
4,53759,Men,Apparel,Topwear,Tshirts,Grey,Summer,2012.0,Casual,Puma Men Grey T-shirt


3. Basic information

In [44]:
print("Dataset shape:", df.shape)
df.info()

Dataset shape: (44446, 10)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44446 entries, 0 to 44445
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  44446 non-null  int64  
 1   gender              44446 non-null  object 
 2   masterCategory      44446 non-null  object 
 3   subCategory         44446 non-null  object 
 4   articleType         44446 non-null  object 
 5   baseColour          44431 non-null  object 
 6   season              44425 non-null  object 
 7   year                44445 non-null  float64
 8   usage               44129 non-null  object 
 9   productDisplayName  44439 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 3.4+ MB


In [45]:
df.describe(include='all')

,id,gender,masterCategory,subCategory,articleType,baseColour,season,year,usage,productDisplayName
count,44446.000000,44446,44446,44446,44446,44431,44425,44445.000000,44129,44439
unique,NaN,5,7,45,143,46,4,NaN,8,31136
top,NaN,Men,Apparel,Topwear,Tshirts,Black,Summer,NaN,Casual,Lucera Women Silver Earrings
freq,NaN,22165,21400,15405,7070,9732,21476,NaN,34414,82
mean,29692.631350,NaN,NaN,NaN,NaN,NaN,NaN,2012.805940,NaN,NaN
std,17048.234982,NaN,NaN,NaN,NaN,NaN,NaN,2.126401,NaN,NaN
min,1163.000000,NaN,NaN,NaN,NaN,NaN,NaN,2007.000000,NaN,NaN
25%,14770.250000,NaN,NaN,NaN,NaN,NaN,NaN,2011.000000,NaN,NaN
50%,28609.500000,NaN,NaN,NaN,NaN,NaN,NaN,2012.000000,NaN,NaN
75%,44678.750000,NaN,NaN,NaN,NaN,NaN,NaN,2015.000000,NaN,NaN


4. Check missing values

In [46]:
df.isnull().sum()

id                      0
gender                  0
masterCategory          0
subCategory             0
articleType             0
baseColour             15
season                 21
year                    1
usage                 317
productDisplayName      7
dtype: int64

In [47]:
(df.isnull().mean() * 100).sort_values(ascending=False)

usage                 0.713225
season                0.047248
baseColour            0.033749
productDisplayName    0.015749
year                  0.002250
id                    0.000000
gender                0.000000
masterCategory        0.000000
subCategory           0.000000
articleType           0.000000
dtype: float64

6. Check the categories

In [48]:
print("Master Categories:")
print(df["masterCategory"].value_counts())

Master Categories:
masterCategory
Apparel           21400
Accessories       11289
Footwear           9222
Personal Care      2404
Free Items          105
Sporting Goods       25
Home                  1
Name: count, dtype: int64


In [49]:
print("\nSub Categories:")
print(df["subCategory"].value_counts().head(20))


Sub Categories:
subCategory
Topwear                     15405
Shoes                        7344
Bags                         3055
Bottomwear                   2694
Watches                      2542
Innerwear                    1808
Jewellery                    1080
Eyewear                      1073
Fragrance                    1012
Sandal                        963
Wallets                       933
Flip Flops                    915
Belts                         811
Socks                         698
Lips                          527
Dress                         478
Loungewear and Nightwear      470
Saree                         427
Nails                         329
Makeup                        307
Name: count, dtype: int64


In [50]:
print("\nArticle Types:")
print(df["articleType"].value_counts().head(30))


Article Types:
articleType
Tshirts                  7070
Shirts                   3217
Casual Shoes             2846
Watches                  2542
Sports Shoes             2036
Kurtas                   1844
Tops                     1762
Handbags                 1759
Heels                    1323
Sunglasses               1073
Wallets                   936
Flip Flops                916
Sandals                   897
Briefs                    849
Belts                     813
Backpacks                 724
Socks                     686
Formal Shoes              637
Perfume and Body Mist     614
Jeans                     609
Shorts                    547
Trousers                  530
Flats                     500
Bra                       477
Dresses                   464
Sarees                    427
Earrings                  417
Deodorant                 347
Nail Polish               329
Lipstick                  315
Name: count, dtype: int64


7. Check product text

In [51]:
print("Missing product names:", df["productDisplayName"].isnull().sum())

print("Unique product names:", df["productDisplayName"].nunique())

print("Duplicate product names:",
      df["productDisplayName"].duplicated().sum())

Missing product names: 7
Unique product names: 31136
Duplicate product names: 13309


8. Verify image availability

In [52]:
image_dir = DATA_DIR /"images"
image_files = list(image_dir.glob("*.jpg"))
print("Images found:", len(image_files))
print("csv rows:", len(df))

Images found: 44441
csv rows: 44446


In [54]:
image_ids = {
    int(path.stem)
    for path in image_files
}
df["image_exists"] = df["id"].isin(image_ids)
print(df["image_exists"].value_counts())

image_exists
True     44441
False        5
Name: count, dtype: int64


In [55]:
print("Missing images", (~df["image_exists"]).sum())

Missing images 5


9. Load cleaned dataset

In [56]:
CLEAN_DATA_PATH = Path("../data/processed/styles_clean.csv")
df_clean = pd.read_csv(CLEAN_DATA_PATH)
print("Shape:", df_clean.shape)
df_clean.head()

Shape: (44446, 10)


,id,gender,masterCategory,subCategory,articleType,baseColour,season,year,usage,productDisplayName
0,15970,Men,Apparel,Topwear,Shirts,Navy Blue,Fall,2011.0,Casual,Turtle Check Men Navy Blue Shirt
1,39386,Men,Apparel,Bottomwear,Jeans,Blue,Summer,2012.0,Casual,Peter England Men Party Blue Jeans
2,59263,Women,Accessories,Watches,Watches,Silver,Winter,2016.0,Casual,Titan Women Silver Watch
3,21379,Men,Apparel,Bottomwear,Track Pants,Black,Fall,2011.0,Casual,Manchester United Men Solid Black Track Pants
4,53759,Men,Apparel,Topwear,Tshirts,Grey,Summer,2012.0,Casual,Puma Men Grey T-shirt


10. Check all subCategory classes

In [57]:
df_clean["subCategory"].value_counts()

subCategory
Topwear                     15405
Shoes                        7344
Bags                         3055
Bottomwear                   2694
Watches                      2542
Innerwear                    1808
Jewellery                    1080
Eyewear                      1073
Fragrance                    1012
Sandal                        963
Wallets                       933
Flip Flops                    915
Belts                         811
Socks                         698
Lips                          527
Dress                         478
Loungewear and Nightwear      470
Saree                         427
Nails                         329
Makeup                        307
Headwear                      293
Ties                          258
Accessories                   143
Scarves                       118
Cufflinks                     108
Apparel Set                   106
Free Gifts                    104
Stoles                         90
Skin Care                      77
Sk

11. Check number of unique classes

In [58]:
print("Unique master categories:", df_clean["masterCategory"].nunique())
print("Unique sub categories:", df_clean["subCategory"].nunique())
print("Unique article types:", df_clean["articleType"].nunique())

Unique master categories: 7
Unique sub categories: 45
Unique article types: 143


12. Check smallest classes

In [59]:
print("Smallest subCategory classes:")
print(df_clean["subCategory"].value_counts().sort_values().head(15))

Smallest subCategory classes:
subCategory
Vouchers               1
Home Furnishing        1
Sports Accessories     3
Wristbands             4
Beauty Accessories     4
Umbrellas              6
Perfumes               6
Water Bottle           7
Bath and Body         12
Hair                  19
Gloves                20
Sports Equipment      21
Shoe Accessories      24
Mufflers              38
Eyes                  43
Name: count, dtype: int64


In [60]:
print("Smallest articleType classes:")
print(df_clean["articleType"].value_counts().sort_values().head(20))

Smallest articleType classes:
articleType
Ipad                   1
Hair Accessory         1
Cushion Covers         1
Suits                  1
Mens Grooming Kit      1
Shoe Laces             1
Body Wash and Scrub    1
Rain Trousers          2
Face Serum and Gel     2
Ties and Cufflinks     2
Key chain              2
Lounge Tshirts         3
Tablet Sleeve          3
Trolley Bag            3
Hat                    3
Lip Plumper            4
Lehenga Choli          4
Beauty Accessory       4
Robe                   4
Makeup Remover         4
Name: count, dtype: int64


13. Check cleaned dataset missing values

In [61]:
df_clean.isnull().sum()

id                      0
gender                  0
masterCategory          0
subCategory             0
articleType             0
baseColour             15
season                 21
year                    1
usage                 317
productDisplayName      7
dtype: int64

14. 20-class selection

In [62]:
TARGET_CLASSES = [
    "Topwear",
    "Shoes",
    "Bags",
    "Bottomwear",
    "Watches",
    "Innerwear",
    "Jewellery",
    "Eyewear",
    "Fragrance",
    "Sandal",
    "Wallets",
    "Flip Flops",
    "Belts",
    "Socks",
    "Lips",
    "Dress",
    "Loungewear and Nightwear",
    "Saree",
    "Nails",
    "Makeup"
]

df_model = df_clean[
    df_clean["subCategory"].isin(TARGET_CLASSES)
].copy()

print("Filtered dataset shape:", df_model.shape)

print("\nNumber of classes:")
print(df_model["subCategory"].nunique())

print("\nClass distribution:")
print(df_model["subCategory"].value_counts())

Filtered dataset shape: (42871, 10)

Number of classes:
20

Class distribution:
subCategory
Topwear                     15405
Shoes                        7344
Bags                         3055
Bottomwear                   2694
Watches                      2542
Innerwear                    1808
Jewellery                    1080
Eyewear                      1073
Fragrance                    1012
Sandal                        963
Wallets                       933
Flip Flops                    915
Belts                         811
Socks                         698
Lips                          527
Dress                         478
Loungewear and Nightwear      470
Saree                         427
Nails                         329
Makeup                        307
Name: count, dtype: int64


In [64]:
print("Missing productDisplayName:", df_model["productDisplayName"].isnull().sum())
print("Duplicate IDs:", df_model["id"].duplicated().sum())

Missing productDisplayName: 7
Duplicate IDs: 0


In [65]:
image_ids = {
    int(path.stem)
    for path in (DATA_DIR / "images").glob("*.jpg")
}
df_model["image_exists"] = df_model["id"].isin(image_ids)
print(df_model["image_exists"].value_counts())

image_exists
True     42866
False        5
Name: count, dtype: int64


In [66]:
print("Missing images:", (~df_model["image_exists"]).sum())

Missing images: 5


### Target Selection & Modeling Scope ✅

- Selected `subCategory` as the target variable
- Selected 20 sufficiently represented classes
- Excluded extremely rare categories
- Final filtered dataset: 42,871 records
- Confirmed 20 target classes
- Confirmed no duplicate product IDs
- Identified 7 missing product names
- Identified 5 missing images
- Defined multimodal input:
  - Product image
  - Product display name
- Defined classification output:
  - 20 product subcategories


15. Remove incomplete multimodal records

In [67]:
df_model = df_clean[
    df_clean["subCategory"].isin(TARGET_CLASSES)
].copy()
print("Before removing incomplete records:", len(df_model))
df_model = df_model.dropna(
    subset=["productDisplayName"]
).copy()

df_model = df_model[
    df_model["id"].isin(image_ids)
].copy()
print("After removing incomplete records:", len(df_model))
print("Missing text:", df_model["productDisplayName"].isnull().sum())
print("Missing images:", (~df_model["id"].isin(image_ids)).sum())

Before removing incomplete records: 42871
After removing incomplete records: 42859
Missing text: 0
Missing images: 0


16. Create Image Path

In [68]:
IMAGE_DIR = DATA_DIR / "images"

df_model["image_path"] = df_model["id"].apply(
    lambda x: str(IMAGE_DIR / f"{int(x)}.jpg")
)

In [72]:
df_model[["id", "productDisplayName", "subCategory", "image_path"]].head()

,id,productDisplayName,subCategory,image_path
0,15970,Turtle Check Men Navy Blue Shirt,Topwear,../data/raw/fashion-product-images-small/image...
1,39386,Peter England Men Party Blue Jeans,Bottomwear,../data/raw/fashion-product-images-small/image...
2,59263,Titan Women Silver Watch,Watches,../data/raw/fashion-product-images-small/image...
3,21379,Manchester United Men Solid Black Track Pants,Bottomwear,../data/raw/fashion-product-images-small/image...
4,53759,Puma Men Grey T-shirt,Topwear,../data/raw/fashion-product-images-small/image...


In [78]:
print("Invalid image paths:", (~df_model["image_path"].apply(os.path.exists)).sum())

Invalid image paths: 0


17. Create numeric labels

In [79]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
df_model["label"] = label_encoder.fit_transform(df_model["subCategory"])

In [84]:
label_mapping = dict(
    zip(
        label_encoder.classes_,
        label_encoder.transform(label_encoder.classes_)
    )
)
label_mapping = {
    label: int(index)
    for label, index in label_mapping.items()
}
label_mapping

{'Bags': 0,
 'Belts': 1,
 'Bottomwear': 2,
 'Dress': 3,
 'Eyewear': 4,
 'Flip Flops': 5,
 'Fragrance': 6,
 'Innerwear': 7,
 'Jewellery': 8,
 'Lips': 9,
 'Loungewear and Nightwear': 10,
 'Makeup': 11,
 'Nails': 12,
 'Sandal': 13,
 'Saree': 14,
 'Shoes': 15,
 'Socks': 16,
 'Topwear': 17,
 'Wallets': 18,
 'Watches': 19}

18. Save the class mapping

In [85]:
import json

with open("../data/processed/label_mapping.json", "w") as file:
    json.dump(label_mapping, file, indent=4)

In [86]:
df_model.head()

,id,gender,masterCategory,subCategory,articleType,baseColour,season,year,usage,productDisplayName,image_path,label
0,15970,Men,Apparel,Topwear,Shirts,Navy Blue,Fall,2011.0,Casual,Turtle Check Men Navy Blue Shirt,../data/raw/fashion-product-images-small/image...,17
1,39386,Men,Apparel,Bottomwear,Jeans,Blue,Summer,2012.0,Casual,Peter England Men Party Blue Jeans,../data/raw/fashion-product-images-small/image...,2
2,59263,Women,Accessories,Watches,Watches,Silver,Winter,2016.0,Casual,Titan Women Silver Watch,../data/raw/fashion-product-images-small/image...,19
3,21379,Men,Apparel,Bottomwear,Track Pants,Black,Fall,2011.0,Casual,Manchester United Men Solid Black Track Pants,../data/raw/fashion-product-images-small/image...,2
4,53759,Men,Apparel,Topwear,Tshirts,Grey,Summer,2012.0,Casual,Puma Men Grey T-shirt,../data/raw/fashion-product-images-small/image...,17


19. Keep only the columns we will need

In [91]:
df_final = df_model[
    [
        "id",
        "image_path",
        "productDisplayName",
        "subCategory",
        "label"
    ]
].copy()

In [92]:
df_final.head()

,id,image_path,productDisplayName,subCategory,label
0,15970,../data/raw/fashion-product-images-small/image...,Turtle Check Men Navy Blue Shirt,Topwear,17
1,39386,../data/raw/fashion-product-images-small/image...,Peter England Men Party Blue Jeans,Bottomwear,2
2,59263,../data/raw/fashion-product-images-small/image...,Titan Women Silver Watch,Watches,19
3,21379,../data/raw/fashion-product-images-small/image...,Manchester United Men Solid Black Track Pants,Bottomwear,2
4,53759,../data/raw/fashion-product-images-small/image...,Puma Men Grey T-shirt,Topwear,17


In [93]:
df_final.shape

(42859, 5)

In [94]:
df_final["subCategory"].value_counts()

subCategory
Topwear                     15401
Shoes                        7344
Bags                         3053
Bottomwear                   2693
Watches                      2542
Innerwear                    1808
Jewellery                    1080
Eyewear                      1073
Fragrance                    1007
Sandal                        963
Wallets                       933
Flip Flops                    915
Belts                         811
Socks                         698
Lips                          527
Dress                         478
Loungewear and Nightwear      470
Saree                         427
Nails                         329
Makeup                        307
Name: count, dtype: int64

20. Save the modeling dataset

In [95]:
FINAL_DATA_PATH = "../data/processed/model_dataset.csv"

df_final.to_csv(
    FINAL_DATA_PATH,
    index=False
)
print(f"Saved to: {FINAL_DATA_PATH}")

Saved to: ../data/processed/model_dataset.csv


### Final Modeling Dataset 

- Removed records with missing product text
- Removed records with missing images
- Created `image_path`
- Verified all image paths
- Created numerical class labels
- Created `label_mapping.json`
- Created `model_dataset.csv`
- Final dataset contains 42,859 products
- Final dataset contains 20 classes
- Invalid image paths: 0


21. Check duplicate product names

In [97]:
duplicate_test_count = (
    df_final["productDisplayName"]
    .duplicated()
    .sum()
)
print("Duplicate product names:", duplicate_test_count)

Duplicate product names: 12872


In [98]:
unique_text_count = (
    df_final["productDisplayName"]
    .nunique()
)
print("Unique product names:", unique_text_count)

Unique product names: 29987


In [99]:
text_label_groups = (
    df_final
    .groupby("productDisplayName")["subCategory"]
    .nunique()
)
print(
    "Product names appearing with multiple labels:",
    (text_label_groups > 1).sum()
)

Product names appearing with multiple labels: 14


In [100]:
print(
    text_label_groups
    .value_counts()
    .sort_index()
)

subCategory
1    29973
2       14
Name: count, dtype: int64


22. Remove ambiguous product names

In [105]:
df_final.shape

(42859, 5)

In [ ]:
ambiguous_names = text_label_groups[
    text_label_groups > 1
].index

print("Ambiguous product names:", len(ambiguous_names))

df_split = df_final[
    ~df_final["productDisplayName"].isin(ambiguous_names)
].copy()

print("Dataset after removing ambiguous names:", df_split.shape)

Ambiguous product names: 14
Dataset after removing ambiguous names: (42795, 5)


In [107]:
check_groups = (
    df_split.groupby("productDisplayName")["subCategory"].nunique()
)
print(
    "Product names with multiple labels:",
    (check_groups > 1).sum()
)

Product names with multiple labels: 0


23. Create grouped split

In [111]:
from sklearn.model_selection import GroupShuffleSplit

gss = GroupShuffleSplit(
    n_splits = 1,
    test_size = 0.30,
    random_state = 42
)

train_idx, temp_idx = next(
    gss.split(df_split, groups = df_split["productDisplayName"])
)

train_df = df_split.iloc[train_idx].copy()
temp_df = df_split.iloc[temp_idx].copy()
print("Train:", train_df.shape)
print("Temp:", temp_df.shape)

Train: (29901, 5)
Temp: (12894, 5)


In [112]:
gss_test = GroupShuffleSplit(
    n_splits =1,
    test_size = 0.50,
    random_state = 42
)

val_idx, test_idx = next(
    gss_test.split(temp_df, groups = temp_df["productDisplayName"])
)

val_df = temp_df.iloc[val_idx].copy()
test_df = temp_df.iloc[test_idx].copy()

print("Train:", len(train_df))
print("Validation:", len(val_df))
print("Test:", len(test_df))

Train: 29901
Validation: 6435
Test: 6459


In [113]:
train_names = set(train_df["productDisplayName"])
val_names = set(val_df["productDisplayName"])
test_names = set(test_df["productDisplayName"])
print("train n Validation:", len(train_names & val_names))
print("train n Test:", len(train_names & test_names))
print("Validation n Test:", len(val_names & test_names))

train n Validation: 0
train n Test: 0
Validation n Test: 0


In [114]:
print("TRAIN")
print(train_df["subCategory"].value_counts())
print("VALIDATION")
print(val_df["subCategory"].value_counts())
print("TEST")
print(test_df["subCategory"].value_counts())

TRAIN
subCategory
Topwear                     10715
Shoes                        5065
Bags                         2118
Bottomwear                   1867
Watches                      1792
Innerwear                    1297
Eyewear                       749
Jewellery                     747
Fragrance                     701
Sandal                        649
Flip Flops                    648
Wallets                       635
Belts                         567
Socks                         529
Lips                          378
Dress                         367
Loungewear and Nightwear      343
Saree                         268
Nails                         242
Makeup                        224
Name: count, dtype: int64
VALIDATION
subCategory
Topwear                     2425
Shoes                       1038
Bags                         483
Bottomwear                   446
Watches                      358
Innerwear                    243
Eyewear                      190
Sandal                

In [115]:
train_classes = set(train_df["subCategory"])
val_classes = set(val_df["subCategory"])
test_classes = set(test_df["subCategory"])
print("Classes missing from train:", set(TARGET_CLASSES) - train_classes)
print("Classes missing from validation:", set(TARGET_CLASSES) - val_classes)
print("Classes missing from test:", set(TARGET_CLASSES) - test_classes)

Classes missing from train: set()
Classes missing from validation: set()
Classes missing from test: set()


### Leakage-Safe Dataset Split 

- Identified 14 ambiguous product names
- Removed all records associated with ambiguous names
- Final modeling dataset: 42,795 records
- Grouped split using `productDisplayName`
- Train: 29,901 records
- Validation: 6,435 records
- Test: 6,459 records
- Verified zero product-name overlap between splits
- Verified all 20 classes exist in train/validation/test


In [116]:
del temp_df

In [117]:
train_df["split"] = "train"
val_df["split"] = "validation"  
test_df["split"] = "test"

24. Save the three Datasets

In [118]:
PROCESSED_DIR = Path("../data/processed")

train_df.to_csv(
    PROCESSED_DIR / "train.csv",
    index=False
)
val_df.to_csv(
    PROCESSED_DIR / "validation.csv",
    index=False
)
test_df.to_csv(
    PROCESSED_DIR / "test.csv",
    index=False
)

In [119]:
print("Train:", (PROCESSED_DIR / "train.csv").exists()) 
print("Validation:", (PROCESSED_DIR / "validation.csv").exists())
print("Test:", (PROCESSED_DIR / "test.csv").exists())

Train: True
Validation: True
Test: True


In [120]:
train_check = pd.read_csv(PROCESSED_DIR / "train.csv")
val_check = pd.read_csv(PROCESSED_DIR / "validation.csv")
test_check = pd.read_csv(PROCESSED_DIR / "test.csv")

print("Train shape:", train_check.shape)
print("Validation shape:", val_check.shape)
print("Test shape:", test_check.shape)

Train shape: (29901, 6)
Validation shape: (6435, 6)
Test shape: (6459, 6)


In [121]:
print("Train classes:", train_check["subCategory"].nunique())
print("Validation classes:", val_check["subCategory"].nunique())   
print("Test classes:", test_check["subCategory"].nunique())

Train classes: 20
Validation classes: 20
Test classes: 20


## Phase 1 — Project Planning & Dataset Preparation 

### Dataset
- Fashion Product Images (Small)
- 44K+ original products
- Original images kept under `data/raw/`

### Cleaning
- Created `styles_clean.csv`
- Removed incomplete records for multimodal modeling

### Target
- Target column: `subCategory`
- Selected 20 well-represented classes

### Final Dataset
- 42,795 usable products
- 20 classes
- 0 duplicate IDs
- 0 missing images
- 0 missing product names

### Leakage Prevention
- Identified duplicate product names
- Removed ambiguous product names
- Grouped data by `productDisplayName`
- Verified zero text overlap between splits

### Final Splits
- Train: 29,901
- Validation: 6,435
- Test: 6,459

### Saved Files
- `styles_clean.csv`
- `model_dataset.csv`
- `label_mapping.json`
- `train.csv`
- `validation.csv`
- `test.csv`
